### ISSUE TITLE
False error being identified in async code

### ISSUE THREAD

**Issue by IsaacsLab42** [microsoft/pyright#9969]
## Environment data

-   Pylance version: 2025.2.1
-   OS and version: Debian 12 (bookworm)
-   Python version: 3.11.2

## Code Snippet

I am working on learning the Python Niquests library. I copy/pasted a [code snippet](https://niquests.readthedocs.io/en/latest/user/quickstart.html#asyncresponse-for-streams) directly from their website into a Python source file. It runs perfectly fine with no error. However, Pylance highlights the code as having a problem.

Here is the code:

```python
import niquests
import asyncio

async def main() -> None:
    async with niquests.AsyncSession() as s:
        r = await s.get("https://httpbingo.org/get", stream=True)

        async for chunk in await r.iter_content(16):
            print(chunk)

if __name__ == "__main__":
    asyncio.run(main())
```

Here's a screenshot of Pylance showing an error:

![Image](https://github.com/user-attachments/assets/7f02dcb2-ec94-4711-9f63-6fc499f4c384)

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release of pyright.

This issue stems from the fact that the `niquests` library is relying on behavior that is unspecified in the Python typing spec. The `iter_content` method is declared as follows:

```python
    @typing.overload  # type: ignore[override]
    async def iter_content(
        self, chunk_size: int = ..., decode_unicode: Literal[False] = ...
    ) -> typing.AsyncGenerator[bytes, None]: ...

    @typing.overload  # type: ignore[override]
    async def iter_content(
        self, chunk_size: int = ..., *, decode_unicode: Literal[True]
    ) -> typing.AsyncGenerator[str, None]: ...

    async def iter_content(  # type: ignore[override]
        self, chunk_size: int = ITER_CHUNK_SIZE, decode_unicode: bool = False
    ) -> typing.AsyncGenerator[bytes | str, None]:
        ...
        yield
```

Normally, the absence of a `yield` statement in an async function or method causes a type checker to assume that its return type should be wrapped in a `Coroutine` or `CoroutineType`. In this case, the implementation does have a `yield` statement but the overloads do not. Pyright previously had some logic in place to _not_ wrap an overload in `Coroutine` if it was declared in a stub or a protocol definition but had a return type of `AsyncGenerator`. The assumption was that the absence of a `yield` statement in the overload was not intentional on the part of the author and should not be interpreted in this case as implying a `Coroutine`. Mypy doesn't have this logic, so it dutifully wraps the return type with `Coroutine` in this instance. 

The `niquests` library arguably has a bug here. It should be specifying a return type of `Coroutine[Any, Any, AsyncGenerator]` rather than `AsyncGenerator`. This would avoid hitting this ambiguous part of the typing spec.

The typing spec is also partly at fault here because it doesn't clearly specify how type checkers should behave in this circumstance.

I've removed pyright's special-case logic so it more closely matches mypy's in this case.

**Comment by IsaacsLab42** [comment]
Thank you! After a bit more research I found other people have mentioned this issue on the Niquests GitHub page. So, I'm cross-referencing these two issues. https://github.com/jawah/niquests/issues/185.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.395.
